source/slang/glsl.meta.slang

```diff
@@ -2825,21 +2825,1985 @@ public uint rayQueryGetIntersectionTypeEXT(rayQueryEXT q, bool committed)
     return 0;
 }
 
+// TODO: implementation of built-in variables; proper tests; these are stubs
+// likley related to the following issue since GLSL adds new 
+// 'system' variables: https://github.com/shader-slang/slang/issues/411
 
-//
-// Subgroup
-//
+#define glsl_caps_shader_subgroup(param1) __glsl_extension(param1) [require(glsl)] 
+#define spirv_caps_shader_subgroup(param1) __spirv_version(param1) [require(spirv)] 
+
+//macros used to make implementation and modification reasonable under regex
+#define macroAnyVecGenericBuiltIn __generic<T : __BuiltinType, let N : int>
+#define macroAnyVecGenericLogical __generic<T : __BuiltinLogicalType, let N : int>
+#define macroAnyVecGenericArithmetic __generic<T : __BuiltinArithmeticType, let N : int>
+#define macroAnyVec vector<T, N>
+
+__generic<T : __BuiltinType>
+void shader_subgroup_preamble() {
+    // checks needed for shader_subgroup functions; __requireGLSLExtension does not work, the call fails to find the OP
+    //__target_switch
+    //{
+    //case glsl:
+        //if (__type_equals<T, half>()) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_float16");
+        //if (__type_equals<T, uint8_t>()) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int8");
+        //if (__type_equals<T, uint16_t>()) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int16");
+        //if (__type_equals<T, uint64_t>()) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int64");
+        //return;
+    //}
+    return;
+} 
+
+// GL_KHR_shader_subgroup_basic Built-in Variables 
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_NumSubgroups;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupID;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupSize;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupInvocationID;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupEqMask;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupGeMask;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupGtMask;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupLeMask;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupLtMask;
```

Comment:
Shouldn't they have "in" or "out" modifiers?

Reply:
this is not permitted syntax as per Slang, it will cause a compile error

built-ins must either be defined through their internal c++ side or inheriting from the C++ side's interface of a built-in